### ISSUE TITLE
acid3 uses unpaired surrogates

### ISSUE THREAD

**Issue by jruderman** [servo/servo#6519]
Visiting http://acid3.acidtests.org/

thread 'ScriptTask PipelineId(0)' panicked at 'called `Result::unwrap()` on an `Err` value: FromUtf16Error(())', src/libcore/result.rs:731
    stack backtrace:
       1:        0x107894065 - sys::backtrace::write::hab580a18e6ade3dauns
       2:        0x107897685 - panicking::on_panic::h1844ce5c03c54289BFw
       3:        0x1078851a2 - rt::unwind::begin_unwind_inner::hbe10c68a88d62d39Nnw
       4:        0x107885809 - rt::unwind::begin_unwind_fmt::h1cd2a6c7dee6575eTmw
       5:        0x10789705c - rust_begin_unwind
       6:        0x1078bfec5 - panicking::panic_fmt::h2f556e59c40f3e3cE8B
       7:        0x1067c7182 - dom::bindings::conversions::jsstring_to_str::h5792c886e1f6e49fXtd
       8:        0x1067cbc69 - dom::bindings::conversions::DOMString.FromJSValConvertible::from_jsval::h1afc9893bea3bbdbCxd
       9:        0x1068f9588 - dom::bindings::codegen::Bindings::HTMLInputElementBinding::set_value::h50e535774fb4155dyF5
      10:        0x1078c8b1a - CallJitSetterOp
      11:        0x1067c8652 - dom::bindings::utils::call_setter::hed79a0ad67ff6e85xrc
      12:        0x1067c839a - dom::bindings::utils::generic_call::h3ebb307afb54a3a8bnc
      13:        0x106d17824 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
      14:        0x106d0fa17 - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
      15:        0x106d2d38e - _ZN2js20InvokeGetterOrSetterEP9JSContextP8JSObjectN2JS5ValueEjPS5_NS4_13MutableHandleIS5_EE
      16:        0x106d49ecb - _ZN2js17NativeSetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_IP8JSObjectEENS3_I4jsidEENS_13QualifiedBoolENS2_13MutableHandleINS2_5ValueEEERNS2_14ObjectOpResultE
      17:        0x106d262d5 - _ZL9InterpretP9JSContextRN2js8RunStateE
      18:        0x106d1f059 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
      19:        0x106d17a19 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
      20:        0x106d0fa17 - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
      21:        0x106e669b1 - _ZN2js3jitL14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_15ICCall_FallbackEjPN2JS5ValueENS7_13MutableHandleIS8_EE

**Comment by jdm** [comment]
We apparently get non-latin1 characters that are not valid utf16.

**Comment by boghison** [comment]
What actually causes this error? I'd like to investigate. Does it happen [here](https://github.com/servo/servo/blob/master/components/script/dom/bindings/conversions.rs#L334)?

**Comment by jdm** [comment]
Yep!

**Comment by boghison** [comment]
@jdm Yes, but the comment says that it fails if it's not valid UTF16, then that function is working properly. Then what is the problem? Is there supposed to be some other expected behavior (like a console error and a null string)?

**Comment by jdm** [comment]
We should figure out what the input is, and what other browsers do with it. It might be coming from test 53, in which case it shouldn't actually be invalid.

**Comment by jdm** [comment]
No wait, it's test 68:
```
      // test 68: UTF-16 surrogate pairs, from David Chan
      // 
      // In The Unicode Standard 5.0, it is explicitly permitted to
      // allow malformed UTF-16, that is, to leave the string alone.
      // (http://www.unicode.org/versions/Unicode5.0.0):
      // 
      // section 2.7: "...strings in ... ECMAScript are Unicode 16-bit
      // strings, but are not necessarily well-formed UTF-16
      // sequences.  In normal processing, it can be far more
      // efficient to allow such strings to contain code unit
      // sequences that are not well-formed UTF-16 -- that is,
      // isolated surrogates"
      // 
      // On the other hand, if the application wishes to ensure
      // well-formed character sequences, it may not permit the
      // malformed sequence and it must regard the first codepoint as
      // an error:
      // 
      // Section 3.2: "C10. When a process interprets a code sequence
      // which purports to be in a Unicode character encoding form, it
      // shall treat ill-formed code unit sequences as an error
      // condition and shall not interpret such sequences as
      // characters.
      // [...]
      // For example, in UTF-8 every code unit of the form 110....2
      // must be followed by a code unit of the form 10......2. A
      // sequence such as 110.....2 0.......2 is ill-formed and must
      // never be generated. When faced with this ill-formed code unit
      // sequence while transforming or interpreting text, a
      // conformant process must treat the first code unit 110.....2
      // as an illegally terminated code unit sequence~Wfor example,
      // by signaling an error, filtering the code unit out, or
      // representing the code unit with a marker such as U+FFFD
      // replacement character."
      // 
      // So it would be permitted to do any of the following:
      // 1) Leave the string alone
      // 2) Remove the unpaired surrogate
      // 3) Replace the unpaired surrogate with U+FFFD
      // 4) Throw an exception
```

**Comment by boghison** [comment]
@jdm So basically it is OK to throw and exception, right?

**Comment by jdm** [comment]
I would be tempted to replace the invalid valid with U+FFFD like http://mxr.mozilla.org/servo/source/components/script/dom/xmlhttprequest.rs#1038 (but following the example at http://doc.servo.org/encoding/index.html labelled "To decode a byte sequence with invalid sequences:", and using the UTF_16LE encoding (http://doc.servo.org/encoding/all/constant.UTF_16LE.html))

**Comment by boghison** [comment]
@jdm OK, I am going to take care of this. Just to know, this should be applied to the actual decoding function (ie not the specific case), right?

**Comment by jdm** [comment]
I believe we want this there, yes, but maybe @SimonSapin can confirm.

**Comment by boghison** [comment]
@jdm I am not sure whether my changes in 1cc1316 actually do what's required. There's no panic, but I changed the parameter in unsafe to be as u8 instead of u16, because that's what the decoding function accepts

**Comment by jdm** [comment]
I'm concerned that the length may be half as long as it should be now, since u16 is two bytes vs. u8. I think ideally we would create a u16 slice as before, then extract a u8 slice of the underlying bytes, but I'm not quite sure how to accomplish that.

**Comment by boghison** [comment]
@jdm Yes, I thought about the length as well, but I just saved an html file as UTF-16 and added the meta tag for the charset, it was working ok

**Comment by jdm** [comment]
I don't think that's testing what you expected. This code is run for strings coming from JS that get passed to the Rust implementation of various DOM APIs. With regards to my previous comment, let's cast the pointer and multiply the length by 2, with an explanatory comment.

**Comment by boghison** [comment]
@jdm I edited the commit, is that what you meant?

**Comment by jdm** [comment]
I'd say "length is the number of u16 elements, while we need the number of u8 bytes" and move it onto the preceding line.

**Comment by Ms2ger** [comment]
The correct solution is to move DOMString to wtf-8, but it's not at all clear to me that making it an alias for USVString is better than throwing or panicking.

**Comment by frewsxcv** [comment]
https://github.com/servo/servo/commit/243446eff33d596b192a7190d5dfe3bb1a6e07dd caused the changes in #6595 to get reverted; should this be reopened?

**Comment by dralley** [comment]
This doesn't crash anymore, if ```-Z replace-surrogates``` is specified

Without it being specified, it still crashes, but with a different traceback.  And it tells you to specify that flag.

**Comment by jdm** [comment]
Fixed by #35381.
